runtime/compiler/env/j9fieldsInfo.cpp

```diff
@@ -166,127 +128,81 @@ TR_VMFieldsInfo::TR_VMFieldsInfo(TR::Compilation * comp, J9Class *aClazz, int bu
          TR_ASSERT_FATAL(supClass, "Found NULL supClass in inheritance chain");
          }
 
-      romCl = supClass->romClass;
-
-      // iterate through the fields creating TR_VMField and inserting them into a List
-
-      currentField = romFieldsStartDo(romCl, &state);
-      while (currentField)
-         {
-         if ((currentField->modifiers & J9AccStatic) == 0)
-            {
-            totalNumFields++;
-            _numRefSlotsInObject += buildField(supClass, currentField);
-            }
-         else
-            {
-            buildField(supClass, currentField);
-            }
-         currentField = romFieldsNextDo(&state);
-         }
+      collectFieldInfo(supClass);
       }
 
    // copy the GCData
-   numSlotsInObject = (aClazz->totalInstanceSize + numBytesInSlot - 1)/numBytesInSlot;
-   numRefs = 0;
-   descriptorPtr = aClazz->instanceDescription;
-
-   // null terminated
-   _gcDescriptor = (int32_t *) _comp->trMemory()->allocateMemory((_numRefSlotsInObject+1)*sizeof(int32_t), allocKind);
-   _gcDescriptor[_numRefSlotsInObject] = 0;
+   UDATA *descriptorPtr = aClazz->instanceDescription;
+   UDATA descriptorWord=0;
 
-   slotsInHeader = (TR::Compiler->om.objectHeaderSizeInBytes()/numBytesInSlot);
-   countSlots = slotsInHeader;
-   bitIndex = 0;
+   int32_t bitIndex = 0;
    if ( ((UDATA) descriptorPtr) & BCT_J9DescriptionImmediate )
       {
       bitIndex++;
       descriptorWord = ((UDATA) descriptorPtr) >> 1;
       }
    else
-     {
-     descriptorWord = descriptorPtr[0];
-     }
+      {
+      descriptorWord = descriptorPtr[0];
+      }
 
+   const int32_t numBytesInSlot = TR::Compiler->om.sizeofReferenceField();
+   const int32_t numBitsInWord = 8*sizeof(decltype(*(aClazz->instanceDescription)));
+   const int32_t numSlotsInObject = (aClazz->totalInstanceSize + numBytesInSlot - 1)/numBytesInSlot;
+   const uintptr_t slotsInHeader = (TR::Compiler->om.objectHeaderSizeInBytes()/numBytesInSlot);
+   uintptr_t countSlots = slotsInHeader;
    while (1)
       {
-       if ( descriptorWord & 0x1 )
-          {
-           _gcDescriptor[numRefs++] = countSlots;
-          }
-       countSlots++;
-       if (countSlots >= (slotsInHeader + numSlotsInObject))
-          {
-             break;
-          }
-       if (bitIndex == (numBitsInWord - 1))
-          {
-          descriptorPtr++;
-          bitIndex = 0;
-          descriptorWord = *descriptorPtr;
-          }
-       else
-          {
-          descriptorWord = descriptorWord >> 1;
-          bitIndex++;
-          }
+      if ( descriptorWord & 0x1 )
```

Comment:
Is there a macro or other helper to do this test and the shifting? Just wanting to avoid magic values if possible...

Reply:
Yeah, I don't like the magic constant either (actually I don't like any of this `descriptorWord` code). But, I couldn't find any macros for it and it looks like the GC also does it this way (although the code is much more elegant):

https://github.com/eclipse/openj9/blob/a91c5af9b85a1894fd641c03956def8d97b0eea4/runtime/gc_base/ClassModel.hpp#L114-L138